Additional seed constraints

badnames opened the pull request:
<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [x] I added a line to [changelog.md](/doc/changelog.md)
- [x] The PR is rebased with current master.
- [x] Details of what you changed are in commit messages.
- [x] References to issues, e.g. `close #X`, are in the commit messages and changelog.
- [x] The buildserver is happy.

<!--
If you have any troubles fulfilling these criteria, please write about the trouble as comment in the PR.
We will help you, but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For documentation fixes, spell checking, and similar none of these points below need to be checked.
Otherwise please check these points when getting a PR done:
-->

- [ ] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [ ] I fully described what my PR does in the documentation
- [ ] I fixed all affected documentation
- [ ] I fixed the introduction tour
- [ ] I wrote migrations in a way that they are compatible with already present data
- [ ] I fixed all affected decisions
- [ ] I added automated tests or a [manual test protocol](../doc/tests/manual/protocol.md)
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](https://github.com/ElektraInitiative/PermaplanT/blob/master/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)
- [ ] Exceptions to any guidelines are documented

## Review

<!--
Reviewers can copy&check the following to their review.
Also the checklist above can be used.
But also the PR creator should check these points when getting a PR done:
-->

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] I've read through the whole documentation
- [ ] I've checked conformity to guidelines

markus2330 commented:
jenkins build please

markus2330 commented on frontend/src/config/i18n/de/seeds.json:
```diff
@@ -33,6 +33,7 @@
 
     "btn_create_seed": "Samen speichern",
     "error_create_seed": "Sorry, ich habe Schwierigkeiten den Samen auf meinem Server abzulegen. Bitte versuche es später erneut.",
+    "error_seed_already_exists": "Ein Samen mit dieser Kombination aus Namen und Pflanze existiert bereits! Bitte ändere die entsprechenden Felder und versuche es erneut.",
```
Good suggestions. Please also have a look at error message guidelines: doc/guidelines/ui.md

They should start with "Sorry, " or "Entschuldigung, " in English and be personalized (the app says that it cannot do something as requested).

markus2330 commented on frontend/src/config/i18n/de/seeds.json:
```diff
@@ -43,7 +44,7 @@
   },
   "harvest_year": "Bezugsjahr",
-  "additional_name": "Zusätzlicher Name",
+  "additional_name": "Zusätzlicher Name (' and - nicht erlaubt)",
```
```suggestion
  "additional_name": "Zusätzlicher Name (- nicht zulässig)",
```

markus2330 commented on frontend/src/config/i18n/en/seeds.json:
```diff
@@ -42,7 +43,7 @@
   },
   "harvest_year": "Harvest Year",
-  "additional_name": "Additional Name",
+  "additional_name": "Additional Name (' and - not allowed)",
```
```suggestion
  "additional_name": "Additional Name (- not allowed)",
```

markus2330 commented on frontend/src/features/seeds/routes/CreateSeed.tsx:
```diff
@@ -23,7 +24,14 @@ export function CreateSeed() {
+    onError: (error) => {
+      const errorTyped = error as AxiosError;
+
+      if (errorTyped.response?.status === 409) {
```
@tschawax great suggestion! :heart: Are they maybe already somewhere defined by Axios?

markus2330 commented on frontend/src/features/seeds/routes/CreateSeed.tsx:
```diff
@@ -23,7 +24,14 @@ export function CreateSeed() {
+      const errorTyped = error as AxiosError;
+
+      if (errorTyped.response?.status === 409) {
+        toast(t('seeds:create_seed_form.error_seed_already_exists'));
```
Can you show this in the respective fields instead of the toast?

badnames replied:
Yes I could, but that might be worse from a usability standpoint than the current solution, since we only use either standard HTML 5 validation or toasts to show errors at the moment.

markus2330 replied:
afaik @Bushuo also used zod which is much more than standard HTML 5 validation?

badnames replied:
So I tried to give feedback through standard HTML validation messages, but encountered a few difficulties.

1. Zod seems to be a client side validation library, but the error that we are displaying here can only be detected on the backend.
2. You can set a custom validation message using a JS-DOM method. 
    However, since we are using react this can only be done using a ref to the input field.
    For some reason getting said ref turned out to be very tricky, and I could not get it to work until now.

markus2330 replied:
Thanks for answering the question. Then let us leave the error currently as it is.

markus2330 reviewed (changes requested):
Great job, some usability problems, though.

markus2330 commented on backend/migrations/2023-09-20-073850_unique_seeds/up.sql:
```diff
@@ -0,0 +1,10 @@
+-- when seeds are displayed with their full name.
+ALTER TABLE seeds
+ADD CONSTRAINT seeds_disalow_confusing_characters
+CHECK (name ~ '^(?!.*(\-|\'')).*$');
```
And regex also seems to fail to check for at least one character or number?

badnames replied:
> But... maybe I'm wrong on this one and the additional single quote is escaping the following single quote 🤔

Yes thats it. Apparently you can escape single quotes in sql by doubling them.

markus2330 replied:
Maybe you can add a small comment there, as there was some confusion about it.

markus2330 commented on frontend/src/features/seeds/components/CreateSeedForm.tsx:
```diff
@@ -180,6 +180,7 @@ const CreateSeedForm = ({
               id="name"
               register={register}
               onChange={onChange}
+              pattern="^(?!.*(-|')).*$"
```
Same here. see #916

markus2330 commented:
jenkins build please

markus2330 commented on frontend/src/config/i18n/de/seeds.json:
```diff
@@ -43,7 +44,7 @@
   },
   "harvest_year": "Bezugsjahr",
-  "additional_name": "Zusätzlicher Name",
+  "additional_name": "Zusätzlicher Name (≥1 Buchstabe oder Zahl; '-' nicht erlaubt)",
```
```suggestion
  "additional_name": "Zusätzlicher Name (min. 1 Zeichen; '-' nicht erlaubt)",
```

markus2330 commented on frontend/src/config/i18n/en/seeds.json:
```diff
@@ -42,7 +43,7 @@
   },
   "harvest_year": "Harvest Year",
-  "additional_name": "Additional Name",
+  "additional_name": "Additional Name (≥1 character or number, '-' not allowed)",
```
```suggestion
  "additional_name": "Additional Name (min.1 character, '-' not allowed)",
```

markus2330 reviewed (changes requested):
descriptive text too long, doesn't fit:

![image](https://github.com/ElektraInitiative/PermaplanT/assets/7949075/1f2a7843-0bd9-48f8-8877-c3ac7235765e)

markus2330 commented:
And please add some tests for what is allowed and what not.

badnames commented:
> `>=1...` is superfluous since it's a mandatory field - we should just make sure to trim the input so it doesn't accept pure whitespaces.

This might not have been communicated clearly enough, but the main idea was to tell the user that are string consisting entirely of special characters (e.g. !"§$%) is not allowed.

>Actually, we shouldn't show such regex descriptions at all in labels. In most cases, it just unnecessarily floods the frontend.
> That description could/should be part of the appearing html5 regex message in case of empty/invalid input. And if we really want to inform the user upfront of disallowed characters, we could use a placeholder text.

Thanks for the feedback.
I have removed the message from the fields label and moved it to the error message.
![image](https://github.com/ElektraInitiative/PermaplanT/assets/27735433/fca8be54-b143-424d-ba55-b2ca6e494e3d)

markus2330 commented on frontend/src/config/i18n/en/seeds.json:
```diff
@@ -43,6 +44,7 @@
   },
   "harvest_year": "Harvest Year",
   "additional_name": "Additional Name",
+  "additional_name_pattern_hint": "At least one letter or number, '-' is not allowed.",
```
```suggestion
  "additional_name_pattern_hint": "At least one letter or number, '-' is not allowed",
```

markus2330 commented on frontend/src/config/i18n/de/seeds.json:
```diff
@@ -44,6 +45,7 @@
   },
   "harvest_year": "Bezugsjahr",
   "additional_name": "Zusätzlicher Name",
+  "additional_name_pattern_hint": "Mindestens ein Buchstabe oder eine Ziffer, '-' ist nicht erlaubt.",
```
```suggestion
  "additional_name_pattern_hint": "Mindestens ein Buchstabe oder eine Ziffer, '-' ist nicht erlaubt",
```

markus2330 reviewed (approved):
Remove double ..

markus2330 commented:
Great job, looks much better this way.
